### PR TITLE
refactor(tool): modularize web search and type result normalization (#291)

### LINKED ISSUE #293: [Feature] Refactor web_search into a typed provider contract and module split
### Problem statement / motivation

**Phase 1 of #291**  
**Scope**: Pure refactor (Zero behavior change, byte-identical output)

Currently, `crates/agentic-server-core/src/tool/web_search.rs` (807 lines) has three architectural limitations that block adding new search providers cleanly:
1. **Untyped Pass-Through to the LLM**:  
   `execute_search` copies raw You.com JSON (`results.web[]` / `results.news[]`) verbatim into the model-facing output string (`web_search.rs:230-236`). `WebSearchProviderResponse` uses untyped `serde_json::Value` (`:262-265`).
2. **Monolithic File**:  
   A single 800+ line file mixes the OpenAI tool adapter, argument validation, and vendor-specific You.com HTTP request/response handling.
3. **Credential Security Leak**:  
   `YouSearchProvider` derives `Debug` over raw `Option<String>`, which prints plaintext API keys in debug logs (`:267-272`).

### Proposed solution

This phase establishes the typed contract and modular layout with **zero external behavior change**, making Phase 2 adding new search provider a pure additive change.

#### A. Modular File Layout
Split `crates/agentic-server-core/src/tool/web_search.rs` into:
```text
crates/agentic-server-core/src/tool/web_search/
├── mod.rs   # WebSearchHandler, WebSearchProvider trait, typed structs, public output mapping
├── args.rs  # WebSearchArguments, typed Freshness, domain post-filter helper
└── you.rs   # YouSearchProvider, HTTP request shaping, and response parsing
```
#### B. Typed Normalization Contract
Replace `serde_json::Value` with strongly typed models:
- **`WebSearchResult`**: Contains `url`, `title`, `description`, `snippets: Vec<String>`, `page_age: Option<String>`, `contents: Option<String>` (livecrawl). Serialized JSON keys remain unchanged; cosmetic fields (thumbnails, favicons) are dropped.
- **`WebSearchProviderMetadata`**: `{ provider, query, latency_ms }`.
- **`WebSearchProviderResponse`**: `{ web: Vec<WebSearchResult>, news: Vec<WebSearchResult>, metadata: WebSearchProviderMetadata }`.
#### C. Configuration & Extension Wiring
- Add `WebSearchProviderKind` (`#[non_exhaustive]`, initially `You` only, with `#[default]`) in `crates/agentic-server-core/src/config.rs`.
- Add `WebSearchProviderConfig.provider`.
- Add `WebSearchHandler::from_config` and hook it into `ToolExecutors::from_config` (`tool/executors.rs:87`).
- **Signature Guarantee**: Keep all existing public constructors (`from_env`, `from_values`, `with_api_key`, `spec_only`) completely unchanged.
- Wrap API credentials in a newtype `ApiKey` with a redacting `Debug` impl (`ApiKey(<redacted>)`).

### Decision on #291  made at the moment

| Question | Default Decision | Rationale |
| :--- | :--- | :--- |
| **Q1 (`page_age` vs `published_at`)** | Keep `page_age` | Guarantees 100% byte-identical serialized output for You.com in Phase 1. |
| **Q4 (Enum Location)** | `config.rs` | Sits naturally next to `SqliteTempStore`, matching repository conventions. |

### Scope Boundaries
| Included in Phase 1  | Explicitly Excluded (Deferred to Phase 2 / Later) |
| :--- | :--- |
| ✅ Module split (`mod.rs`, `args.rs`, `you.rs`) | ❌ No third-party provider code |
| ✅ Typed `WebSearchResult` & `WebSearchProviderResponse` | ❌ No new `config.toml` keys or environment variables |
| ✅ Credential redacting `Debug` | ❌ No default base URL for You.com (#291 Q2) |
| ✅ `WebSearchProviderKind::You` and `from_config` | ❌ `WebSearchProvider` trait remains private |
| ✅ Byte-identical output assertion test | ❌ No doc updates (except `CHANGELOG.md`) |

### Verification Checklist
- [ ] **Byte-Identity Test**: Serialize handler output for existing mock fixture (`tests/web_search_tool_test.rs:291`) and assert exact byte equality with pre-refactor output.
- [ ] **Fixture Test**: Add a redacted real You.com response fixture to verify all non-cosmetic fields are preserved.
- [ ] **Public Output Contract**: Verify `sources_from_output` continues to generate identical `web_search_call.action.sources`.
- [ ] **Integration Tests**: All existing tests pass unmodified across both crates (including all `with_api_key` callers).
- [ ] **Code Hygiene**:
  - `cargo clippy --all-targets -- -D warnings` (clean)
  - `cargo fmt -- --check` (clean)
  - `pre-commit run --all-files` (clean)

### PR DESCRIPTION
**Pull request by Zheng-Lu**
## Summary

Fixes #293. Phase 1 of #291: split web search into argument validation, You.com request/response handling, and the gateway adapter, with a typed provider response contract for future backends.

- Normalize search hits into typed results, retaining modeled content and dropping cosmetic/unknown fields. Field order is deterministic; invalid freshness values fail locally.
- Always emit a per-query metadata object. When You.com omits metadata or returns null, use the submitted query and omit unavailable `search_uuid` and `latency`. This intentionally changes `metadata: [null]` to `metadata: [{"query": "..."}]` in model-facing tool output; it does not add metadata to the public `web_search_call` API.
- Preserve public constructor signatures, the existing `WebSearchProviderConfig` fields, and public source mapping. Add a config constructor and redact credentials in Debug output. Provider selection and per-provider concurrency configuration remain deferred; the provider trait supports a concurrency ceiling.
- Refresh the four Messages gateway-search recordings through the existing recorder against live vLLM using `RedHatAI/Qwen3-Coder-Next-NVFP4`. Keep the original client-tool recordings and document both sets' provenance.

## Test Plan

- `cargo test --workspace`
- `cargo test -p agentic-server --test messages_tool_choice_cassette_test` — all six replay cases pass, including the four previously failing search cases.
- Metadata regression coverage for omitted, null, empty, and null-query provider metadata.
- `cargo clippy --all-targets -- -D warnings`
- `cargo fmt -- --check`
- `uvx pre-commit run --all-files`
- Live recorder validated two upstream requests, one gateway search, and a completed answer for each of the four refreshed recordings.

### CODE REVIEW

File: crates/agentic-server-core/src/config.rs [hunk 1]
```diff
@@ -85,10 +87,70 @@ impl Default for SqliteConfig {
...
+        }
+    }
+}
+
+/// Credentials and limits for the gateway-owned `web_search` provider.
+///
+/// Non-exhaustive so provider options can grow without breaking downstream
+/// construction; build it with [`WebSearchProviderConfig::new`] or
+/// [`Default`] and assign the remaining public fields.
+#[derive(Clone, Default)]
```
Comment (maralbahari) [inline crates/agentic-server-core/src/config.rs:129]: Making the existing public `WebSearchProviderConfig` non-exhaustive and adding required fields breaks downstream struct-literal construction; either treat this as a breaking API change or introduce a separate extensible configuration type.

File: crates/agentic-server-core/tests/web_search_tool_test.rs [hunk 2]
```diff
@@ -462,6 +464,137 @@ async fn web_search_handler_requires_base_url() {
...
+                "query": "rust async",
+                "queries": ["rust async"],
+                "sources": [{"url": "https://example.com/rust", "title": "Rust async guide"}]
+            }
+        })
+    );
+}
+
+/// Synthetic You.com `GET /v1/search` response built from the documented
+/// schema (not a live capture): every documented result and metadata field
```
Comment (maralbahari) [inline crates/agentic-server-core/tests/web_search_tool_test.rs]: This fixture is synthetic, despite the PR claiming a redacted real You.com response; replay a sanitized real provider response so typed normalization is tested against the actual wire format.

**Review by maralbahari — COMMENTED** [review]
Thanks for the refactor job, the design is clean and pluggable pattern makes it easy to support different provider. I have added two inline comments and please resolve the merge conflict with `main` and ensure the existing web-search provider response and aggregate tool-output size limits are preserved.

File: crates/agentic-server-core/src/config.rs (hunk 1 above)
Comment (Zheng-Lu) [reply crates/agentic-server-core/src/config.rs:129]: I've restored `WebSearchProviderConfig` back to its published 0.6.0 shape and kept only the non-breaking improvements:
- A `new()` helper constructor.
- A redacting `Debug` implementation so secrets don't leak into logs.

I will introduce the provider-selection configuration in Phase 2 when Brave Search actually lands.

File: crates/agentic-server-core/tests/web_search_tool_test.rs (hunk 2 above)
Comment (Zheng-Lu) [reply crates/agentic-server-core/tests/web_search_tool_test.rs]: Apologies for the misleading wording earlier. I went back and checked the recorded cassettes under `tests/cassettes/messages_multiround/`, which captured live You.com responses from previous runs. I extracted that live data into our test fixture (only redacting the UUID and rounding the latency).

Testing against the real wire format paid off immediately:
- It revealed that You.com returns `original_thumbnail_url` (which our code was silently dropping, now properly documented and asserted as cosmetic).
- It surfaced a real edge case where You.com omits the `news` key entirely, which our normalization now explicitly tests and handles.

**Review by franciscojavierarceo — APPROVED** [review]
